File: android/src/main/java/com/affise/attribution/react/ResultWrapper.kt
package com.affise.attribution.react

import com.affise.attribution.internal.callback.InternalResult
import com.facebook.react.bridge.Promise

class ResultWrapper(private val promise: Promise) : InternalResult {
    override fun success(data: Any?) {
        promise.resolve(data?.asNativeData())
    }

    override fun error(error: String) {
        promise.reject("affise", error)
    }

    override fun notImplemented() {
        promise.reject("affise", "notImplemented")
    }

    private fun Any.asNativeData() : Any? = when (this) {
        is Map<*, *> -> this.toWritableNativeMap()

        is List<*> -> this.toWritableNativeArray()
        else -> this.toNativeBridgeData()
    }
}

File: ios/ResultWrapper.swift
import React
import AffiseInternal

class ResultWrapper : InternalResult {

    private var resolve: RCTPromiseResolveBlock
    private var reject: RCTPromiseRejectBlock

    public init(_ resolve: @escaping RCTPromiseResolveBlock, _ reject: @escaping RCTPromiseRejectBlock) {
        self.resolve = resolve
        self.reject = reject
    }

    func success(_ result: Any?) {
        resolve(result)
    }

    public func error(_ error: String) {
        reject("affise", error, nil)
    }

    public func notImplemented() {
        reject("affise", "notImplemented", nil)
    }
}
